File: logger/src/test/java/fr/beapp/logger/LoggerTest.java
package fr.beapp.logger;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import org.junit.jupiter.api.Test;

import java.util.LinkedList;
import java.util.List;

import fr.beapp.logger.appender.Appender;
import fr.beapp.logger.appender.DebugAppender;
import fr.beapp.logger.formatter.Formatter;

public class LoggerTest {

    private static class DummyAppender extends Appender {
        public final List<String> acc = new LinkedList<>();

        private DummyAppender(@Logger.LogLevel int level) {
            super(level);
        }

        @Override
        public void log(@Logger.LogLevel int priority, @NonNull String message, @Nullable Throwable tr) {
            acc.add(message);
        }
    }

    private final Formatter formatter = (tr, message, args) -> "formatted: " + message;

    @Test
    public void testLog() {
        DummyAppender appender1 = new DummyAppender(Log.DEBUG);
        DummyAppender appender2 = new DummyAppender(Log.INFO);
        DummyAppender appender3 = new DummyAppender(Log.ASSERT);

        Logger.removeAllAppenders();
        Logger.add(appender1);
        Logger.add(appender2);
        Logger.add(appender3);
        Logger.formatter(formatter);

        Logger.trace("trace");
        Logger.debug("debug");
        Logger.info("info");
        Logger.warn("warn");
        Logger.error("error");

        assertArrayEquals(new String[]{"formatted: debug", "formatted: info", "formatted: warn", "formatted: error"}, appender1.acc.toArray());
        assertArrayEquals(new String[]{"formatted: info", "formatted: warn", "formatted: error"}, appender2.acc.toArray());
        assertArrayEquals(new String[]{}, appender3.acc.toArray());
    }

    @Test
    public void testFindOfType() {
        DummyAppender appender1 = new DummyAppender(Log.INFO);
        DummyAppender appender2 = new DummyAppender(Log.INFO);

        Logger.removeAllAppenders();
        Logger.add(appender1);
        Logger.add(appender2);

        List<DebugAppender> debugAppenders = Logger.findOfType(DebugAppender.class);
        assertEquals(0, debugAppenders.size());

        List<DummyAppender> dummyAppenders = Logger.findOfType(DummyAppender.class);
        assertEquals(2, dummyAppenders.size());
        assertEquals(appender1, dummyAppenders.get(0));
        assertEquals(appender2, dummyAppenders.get(1));
    }

}

File: logger/src/test/kotlin/fr/beapp/logger/LoggerTest.kt
package fr.beapp.logger

import android.util.Log
import fr.beapp.logger.Logger.add
import fr.beapp.logger.Logger.debug
import fr.beapp.logger.Logger.error
import fr.beapp.logger.Logger.filterInstances
import fr.beapp.logger.Logger.findLevelName
import fr.beapp.logger.Logger.findOfType
import fr.beapp.logger.Logger.info
import fr.beapp.logger.Logger.removeAllAppenders
import fr.beapp.logger.Logger.trace
import fr.beapp.logger.Logger.warn
import fr.beapp.logger.Logger.wtf
import fr.beapp.logger.appender.Appender
import fr.beapp.logger.appender.DebugAppender
import fr.beapp.logger.formatter.Formatter
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.LinkedList

class LoggerTest {
	private class DummyAppender(@Logger.LogLevel level: Int) : Appender(level) {
		val acc: MutableList<String> = LinkedList()
		override fun log(@Logger.LogLevel priority: Int, message: String, tr: Throwable?) {
			acc.add(message)
		}
	}

	private val formatter = Formatter { _: Throwable?, message: String?, _: Array<Any?>? -> "formatted: $message" }

	private val silentFormatter = Formatter { _: Throwable?, message: String?, _: Array<Any?>? -> "" }
	private val nullFormatter = Formatter { _: Throwable?, message: String?, _: Array<Any?>? -> null }

	@Test
	fun `test log`() {
		val appender1 = DummyAppender(Log.DEBUG)
		val appender2 = DummyAppender(Log.INFO)
		val appender3 = DummyAppender(Log.ASSERT)
		val appender4 = DummyAppender(Log.VERBOSE)
		removeAllAppenders()
		add(appender1)
		add(appender2)
		add(appender3)
		Logger.formatter = formatter
		trace("trace")
		debug("debug")
		info("info")
		warn("warn")
		error("error")
		warn()
		error()
		wtf("wtf")
		wtf()

		Assertions.assertArrayEquals(
			arrayOf("formatted: debug", "formatted: info", "formatted: warn", "formatted: error", "formatted: null", "formatted: null", "formatted: wtf", "formatted: null"),
			appender1.acc.toTypedArray()
		)
		Assertions.assertArrayEquals(
			arrayOf("formatted: info", "formatted: warn", "formatted: error", "formatted: null", "formatted: null", "formatted: wtf", "formatted: null"),
			appender2.acc.toTypedArray()
		)
		Assertions.assertArrayEquals(arrayOf("formatted: wtf"), appender3.acc.toTypedArray())
		Logger.formatter = silentFormatter
		removeAllAppenders()
		add(appender4)
		trace("trace")
		Assertions.assertArrayEquals(arrayOf<String>(), appender4.acc.toTypedArray())
		Logger.formatter = nullFormatter
		trace("trace")
		Assertions.assertArrayEquals(arrayOf<String>(), appender4.acc.toTypedArray())
	}

	@Test
	fun `test findOfType`() {
		val appender1 = DummyAppender(Log.INFO)
		val appender2 = DummyAppender(Log.INFO)
		removeAllAppenders()
		add(appender1)
		add(appender2)
		val debugAppenders = findOfType(DebugAppender::class.java)
		Assertions.assertEquals(0, debugAppenders.size)
		val dummyAppenders = findOfType(DummyAppender::class.java)
		Assertions.assertEquals(2, dummyAppenders.size)
		Assertions.assertEquals(appender1, dummyAppenders[0])
		Assertions.assertEquals(appender2, dummyAppenders[1])
	}

	@Test
	fun `test filterInstances`() {
		val appender1 = DummyAppender(Log.INFO)
		val appender2 = DummyAppender(Log.INFO)
		removeAllAppenders()
		add(appender1)
		add(appender2)
		val debugAppenders = filterInstances<DebugAppender>()
		Assertions.assertEquals(0, debugAppenders.size)
		val dummyAppenders: List<DummyAppender> = filterInstances()
		Assertions.assertEquals(2, dummyAppenders.size)
		Assertions.assertEquals(appender1, dummyAppenders[0])
		Assertions.assertEquals(appender2, dummyAppenders[1])
	}

	@Test
	fun testFindLevelName() {
		Assertions.assertEquals("TRACE", findLevelName(Log.VERBOSE))
		Assertions.assertEquals("DEBUG", findLevelName(Log.DEBUG))
		Assertions.assertEquals("INFO", findLevelName(Log.INFO))
		Assertions.assertEquals("WARN", findLevelName(Log.WARN))
		Assertions.assertEquals("ERROR", findLevelName(Log.ERROR))
		Assertions.assertEquals("WTF", findLevelName(Log.ASSERT))
		Assertions.assertEquals("", findLevelName(42))
	}


}